internal/pkg/upload/upload.go

```diff
@@ -0,0 +1,341 @@
+	"github.com/gofrs/uuid"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	//these should be configs probably
+	uploadRequestTimeout = time.Hour
+	chunkProgressTimeout = time.Hour / 4
+
+	// specification-designated maximum
+	MaxChunkSize = 4194304 // 4 MiB
+)
+
+var (
+	ErrMaxConcurrentUploads = errors.New("the max number of concurrent uploads has been reached")
+	ErrInvalidUploadID      = errors.New("active upload not found with this ID, it may be expired")
+	ErrFileSizeTooLarge     = errors.New("This file exceeds the maximum allowed file size")
+	ErrMissingChunks        = errors.New("File data incomplete, not all chunks were uploaded")
+	ErrHashMismatch         = errors.New("file integrity hash does not match")
+	//@todo: explicit error for expired uploads
+)
+
+type upload struct {
+	opToken       *throttle.Token
+	chunkThrottle *throttle.Throttle
+	complete      chan<- struct{}
+	chunkRecv     chan<- struct{}
+	begun         bool
+	Info          Info
+}
+
+type Uploader struct {
+	current            map[string]upload
+	mu                 sync.Mutex
+	opThrottle         *throttle.Throttle
+	parallelChunkLimit int
+	sizeLimit          int64
+}
+
+type Info struct {
+	ID        string // upload operation identifier. Ephemeral, just used for the upload process
+	DocID     string // document ID of the uploaded file and chunks
+	Source    string // which integration is performing the upload
+	ChunkSize int64
+	Total     int64
+	Count     int
+	HashSum   string
+	Hasher    hash.Hash
+}
+
+type ChunkInfo struct {
+	ID            int
+	FirstReceived bool
+	Final         bool
+	Upload        Info
+	Token         *throttle.Token
+}
+
+func New(sizeLimit int64, opLimit int, chunkLimit int) *Uploader {
+	return &Uploader{
+		parallelChunkLimit: chunkLimit,
+		sizeLimit:          sizeLimit,
+		opThrottle:         throttle.NewThrottle(opLimit),
+		current:            make(map[string]upload, opLimit),
+	}
+}
+
+// Start an upload operation, as long as the max concurrent has not been reached
+// returns the upload ID
+func (u *Uploader) Begin(size int64, docID string, source string, hashsum string, hasher hash.Hash) (Info, error) {
+	if size <= 0 {
+		return Info{}, errors.New("invalid file size")
+	}
+	if size > u.sizeLimit {
+		return Info{}, ErrFileSizeTooLarge
+	}
+
+	uid, err := uuid.NewV4()
+	if err != nil {
+		return Info{}, fmt.Errorf("unable to generate upload operation ID: %w", err)
+	}
+	id := uid.String()
+
+	token := u.opThrottle.Acquire(id, 300*time.Hour)
+	if token == nil {
+		return Info{}, ErrMaxConcurrentUploads
+	}
+
+	total := time.NewTimer(uploadRequestTimeout)
+	chunkT := time.NewTimer(chunkProgressTimeout)
+	chunkRecv := make(chan struct{})
+	complete := make(chan struct{})
+	// total timer could also be achieved with context deadline and cancelling
+
+	go func() {
+		for {
+			select {
+			case <-total.C: // entire upload operation timed out
+				log.Trace().Str("uploadID", id).Msg("upload operation timed out")
```

Comment:
Is there a way to associate an uploadID back to the agent/endpoint host it originated from? If we run into issues where uploads from particular hosts are timing out, is there a way for us to figure out what hosts those are so we can correlate this to agent/endpoint logs?

Including the agent ID in addition to the uploadID is an easy way to solve this.

Reply:
I see the agent ID included in the request handlers, but given this is asynchronous it looks like we'd have to manually try to map uploadIDs back to log statements for requests which seems tedious.